### PR TITLE
BIG-4194 - add a LRU cache

### PR DESCRIPTION
**Pull request by dlecocq**
@b4hand @arora-richa @neilmb 

This introduces two functionally similar caches -- `AgentCache` and `RobotsCache`. Both implement a LRU of their respective objects (`Agent`s or `Robots`), and honor expiration, subject to a pluggable `ttl` policy.

An exception when fetching is considered to mean that the target site is disallowed, but perhaps at some point that should be made configurable by another policy type or option.

### CODE REVIEW

**Comment by dlecocq** [comment]
At this point, the python 3 builds are failing because of lack of coverage for the fetching (`asis` does not support python 3 yet -- I have a branch where I've been hacking on it, but `bottle`'s python 3 support is suspect).

**Comment by dlecocq** [comment]
I'm reworking this a bit to delegate to a policy about what to do in the case of an exception fetching `robots.txt`. The two main ideas are: 1) cache a disallowed response or 2) cache and reraise the exception.

A common adage is to not cache failures. On the other hand, we've encountered cases with `robots.txt` where the failure comes in the form of a timeout, so if we don't cache timeout responses, every single check to `robots.txt` for a site could take a long time.

This design allows the client to decide what mode of operation is appropriate.

**Comment by dlecocq** [comment]
Rebased, comments addressed.

File: README.md [hunk 1]
```diff
@@ -104,9 +113,49 @@ Robots.robotsUrl('http://userinfo@example.com:8080/path;params?query#fragment')
...
+To this end, the `cache` module provides a notion of a cache policy. It determines what
+to do in the case of an exception. The default is to cache a form of a disallowed response
+for 10 minutes, but you can configure it as you see fit:
+
+```python
+# Do not cache failures (note the `ttl=0`):
+from reppy.cache.policy import ReraiseExceptionPolicy
+cache = AgentCache('my-user-agent', cache_policy=ReraiseExceptionPolicy(ttl=0))
+
+# Cache and reraise failures for 10 minutes (not the `ttl=600`):
```
Comment (arora-richa) [inline README.md]: do you meant `note`?

File: tests/test_cache/test_cache.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        with requests_fixtures('test_robots_allowed'):
+            self.assertFalse(
+                self.cache.allowed('http://example.com/disallowed', 'agent'))
+            self.assertTrue(
+                self.cache.allowed('http://example.com/allowed', 'agent'))
+
+    def test_caches_robots(self):
+        '''Caches robots responses.'''
+        with requests_fixtures('test_caches_robots'):
+            self.cache.get('http://example.com/')
```
Comment (b4hand) [inline tests/test_cache/test_cache.py]: Is there no way to assert this is only called once?

File: README.md (hunk 1 above)
Comment (dlecocq) [inline README.md]: Indeed. I'll change it.

File: tests/test_cache/test_cache.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        with requests_fixtures('test_agent_allowed'):
+            self.assertFalse(
+                self.cache.allowed('http://example.com/disallowed'))
+            self.assertTrue(
+                self.cache.allowed('http://example.com/allowed'))
+
+    def test_caches_agent(self):
+        '''Caches agent responses.'''
+        with requests_fixtures('test_caches_agent'):
+            self.cache.get('http://example.com/')
```
Comment (b4hand) [inline tests/test_cache/test_cache.py]: Ah, that was too subtle for me, but I'm fine with it. Maybe add a comment describing that is the intention.

Comment (dlecocq) [inline tests/test_cache/test_cache.py]: Added a note to that effect

**Pull request merged**